File: 62-Unique_Paths/main.js
/**
 * 自底向上的方案
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function(m, n) {
  let f = Array(n);

  for (let i = 0; i < n; i++) {
    f[i] = Array(m);
  }

  for (let i = 0; i < n; i++) {
    f[i][0] = 1;
  }

  for (let i = 0; i < m; i++) {
    f[0][i] = 1;
  }


  for (let i = 1; i < n; i++) {
    for (let j = 1; j < m; j++) {
      f[i][j] = f[i-1][j] + f[i][j-1];
    }
  }

  return f[n-1][m-1];
};

console.log(uniquePaths(7,3));

File: 62-Unique_Paths/main.go
package main

import (
	"fmt"
)

// 自顶向下的方案
func uniquePaths(m int, n int) int {
	res := make([][]int, m)

	for i := 0; i < m; i++ {
		res[i] = make([]int, n)
	}

	return dp(m-1, n-1, res)
}

// 求解 m*n 方格路径的解法总数
func dp(m, n int, res [][]int) int {
	if m == 0 || n == 0 {
		return 1
	}

	if res[m][n] != 0 {
		return res[m][n]
	}

	res[m][n] = dp(m, n-1, res) + dp(m-1, n, res)
	return res[m][n]
}


func main() {
	fmt.Println(uniquePaths(7, 3))
}